### PR TITLE
[TableGen] Integrate TableGen-based macro fusion

### PR DESCRIPTION
**Pull request by wangpc-pp**
`Fusion` is inherited from `SubtargetFeature` now. Each definition
of `Fusion` will define a `SubtargetFeature` accordingly.

Method `getMacroFusions` is added to `TargetSubtargetInfo`, which
returns a list of `MacroFusionPredTy` that will be evaluated by
MacroFusionMution.

`getMacroFusions` will be auto-generated if the target has `Fusion`
definitions.

### CODE REVIEW

**Comment by wangpc-pp** [comment]
Ping.

**Comment by wangpc-pp** [comment]
Ping.

**Review by arsenm — COMMENTED** [review]
Description says "Future" instead of "Feature"

File: llvm/utils/TableGen/CodeGenTarget.h [hunk 1]
```diff
@@ -149,6 +151,10 @@ class CodeGenTarget {
 
   const CodeGenHwModes &getHwModes() const { return CGH; }
 
+  bool hasMacroFusion() const { return !MacroFusions.empty(); }
```
[the remark this answers is not in the document]
Comment (wangpc-pp) [reply llvm/utils/TableGen/CodeGenTarget.h:154]: This is to simplify `getMacroFusions().empty()`.

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/utils/TableGen/CodeGenTarget.h:154]: There's no reason for it to be virtual then

[the remark this answers is not in the document]
Comment (wangpc-pp) [reply llvm/utils/TableGen/CodeGenTarget.h:154]: Sorry I don't get it. It's not a virtual function. You mean `TargetSubtargetInfo::getMacroFusions()`?

**Comment by wangpc-pp** [comment]
Ping.

File: llvm/include/llvm/CodeGen/TargetSubtargetInfo.h [hunk 2]
```diff
@@ -323,6 +324,12 @@ class TargetSubtargetInfo : public MCSubtargetInfo {
   /// helps removing redundant copies generated by register allocator when
   /// handling complex eviction chains.
   virtual bool enableSpillageCopyElimination() const { return false; }
+
+  /// Enable macro fusion for this subtarget.
+  virtual bool enableMacroFusion() const { return false; }
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:329]: It's virtual here

Comment (wangpc-pp) [reply llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:329]: This function will be overrided in TableGen-generated XXXGenSubtargetInfo classes if subtarget defines some fusions. I think it should be virtual.

Comment (arsenm) [reply llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:329]: Where is the use context where enableMacroFusion is necessary? I assume the use just directly iterates through the returned instances from getMacroFusions. That is, with something like:

```
if (enableMacroFusion()) 
  doStuff(getMacroFusions()) 

```

there's no point in having enableMacroFusion at all

Comment (wangpc-pp) [reply llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:329]: Oh, I get it. The typical usage is like:
```
    if (ST.enableMacroFusion()) {
      ScheduleDAGMI *DAG = createGenericSchedPostRA(C);
      DAG->addMutation(createMacroFusionDAGMutation(ST.getMacroFusions()));
      return DAG;
    }
```
`enableMacroFusion()` is for knowing if we should enable MacroFusion, `getMacroFusions()` returns the predicates.

Yes, I know we can replace `enableMacroFusion()` with `!getMacroFusions().empty()`, but the implementation of `enableMacroFusion()` avoid the creation of `std::vector`.

Let't take RISCV as an example: https://github.com/llvm/llvm-project/pull/72224/commits/1c92927e7d46c2b3c3f173e35afb2eecd4e9ffbf.
`enableMacroFusion()` is a replacement of `RISCVSubtarget::hasMacroFusion()`:
```
bool hasMacroFusion() const {
    return hasLUIADDIFusion() || hasAUIPCADDIFusion() || hasZExtHFusion() ||
           hasZExtWFusion() || hasShiftedZExtWFusion() || hasLDADDFusion();
}
```

Comment (arsenm) [reply llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:329]: But you're just avoiding the creation of an empty vector, which doesn't matter. I would just do
vector Mutations = ST.getMacroFusions()
if (!Mutations.empty())
  DAG->addMutation(createMacroFusion(std::move(Mutations))

Comment (wangpc-pp) [reply llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:329]: OK, I will remove `enableMacroFusion()`.

**Review by arsenm — APPROVED** [review]

**Pull request merged**